### PR TITLE
Include gatk4/postprocessgermlinecnvcalls 

### PR DESCRIPTION
**Pull request by Jorisvansteenbrugge**
This PR adds the final step in the GATK4 germlinecnv caller pipeline: 

> Postprocesses the output of GermlineCNVCaller and generates VCFs and denoised copy ratios

This step was already implemented in a nf-core module: gatk4/postprocessgermlinecnvcalls
 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/createpanelrefs/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/createpanelrefs _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by maxulysse** [comment]
Thanks for the PR @Jorisvansteenbrugge

Don't worry about the failed linting test, #26 will take care of that.

Code looks good to me, I'll ask for @ramprasadn for a complementaty review, as he was the one that implemented part of what you're looking into.

**Comment by maxulysse** [comment]
@nf-core-bot fix linting pretty please :pray:

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 1]
```diff
@@ -167,6 +169,53 @@ workflow GERMLINECNVCALLER_COHORT {
...
+                  .map{data -> [
+                    meta: data[0],
+                    calls: data[1],
+                    model: data[2],
+                    ploidy: data[3]
+                  ]}                
+                  .set{input_channel}
+        
+    GATK4_POSTPROCESSGERMLINECNVCALLS( input_channel )
+
```
Comment (maxulysse) [inline subworkflows/local/germlinecnvcaller_cohort.nf]:
```suggestion
        //
        // create input channels for GATK4_POSTPROCESSGERMLINECNVCALLS
        //
        ch_readcounts_out
            .map{idx -> idx[1].baseName}
            .flatten()
            .map{data -> [id:data]}
            .set{ch_sample_ids}

        GATK4_GERMLINECNVCALLER.out.cohortcalls
            .collect()
            .map{meta, calls1, meta2, calls2 -> [calls1, calls2]}
            .set{cnv_calls}

        GATK4_GERMLINECNVCALLER.out.cohortmodel
            .collect()
            .map{meta, model, meta2, model2 -> [model, model2]}
            .set{cnv_model}

        cnv_calls.concat(cnv_model)
            .collect(flat:false)
            .set{cnv_calls_model}

        GATK4_DETERMINEGERMLINECONTIGPLOIDY.out.calls
            .map{meta, ploidcalls -> ploidcalls}
            .set{ploidy_path}
        cnv_calls
            .concat(cnv_model)
            .concat(ploidy_path)
            .collect(flat:false)
            .set{base_cnv_calls}

        ch_sample_ids.combine(base_cnv_calls)
            .map{data -> [
                meta: data[0],
                calls: data[1],
                model: data[2],
                ploidy: data[3]
            ]}               
            .set{input_channel}

    GATK4_POSTPROCESSGERMLINECNVCALLS( input_channel )
```

Comment (maxulysse) [reply subworkflows/local/germlinecnvcaller_cohort.nf]: Just removed some empty lines and trailing spaces / adjusted indentation for consistency/linting

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 2]
```diff
@@ -181,10 +230,12 @@ workflow GERMLINECNVCALLER_COHORT {
         ch_versions = ch_versions.mix(GATK4_INTERVALLISTTOOLS.out.versions)
         ch_versions = ch_versions.mix(GATK4_DETERMINEGERMLINECONTIGPLOIDY.out.versions)
         ch_versions = ch_versions.mix(GATK4_GERMLINECNVCALLER.out.versions.first())
+        ch_versions = ch_versions.mix(GATK4_POSTPROCESSGERMLINECNVCALLS.out.versions)
 
     emit:
         cnvmodel    = GATK4_GERMLINECNVCALLER.out.cohortmodel
         ploidymodel = GATK4_DETERMINEGERMLINECONTIGPLOIDY.out.model
+        ploidycalls = GATK4_DETERMINEGERMLINECONTIGPLOIDY.out.calls
```
Comment (maxulysse) [inline subworkflows/local/germlinecnvcaller_cohort.nf]:
```suggestion
        ploidycalls = GATK4_DETERMINEGERMLINECONTIGPLOIDY.out.calls
```

File: nextflow.config [hunk 3]
```diff
@@ -29,7 +29,7 @@ params {
     gcnv_model_name             = 'germlinecnvcaller'
     gcnv_padding                = 0
     gcnv_readcount_format       = 'HDF5'
-    gcnv_scatter_content        = 5000
+    gcnv_scatter_content        = 500 // default 5000
```
Comment (maxulysse) [inline nextflow.config]: why changing this value?

File: conf/modules/germlinecnvcaller_cohort.config [hunk 4]
```diff
@@ -87,7 +87,7 @@ process {
         publishDir = [
                 mode: params.publish_dir_mode,
                 path: { "${params.outdir}/germlinecnvcaller/determinegermlinecontigploidy" },
-                pattern: "*-model",
+                pattern: "*-*",
```
Comment (maxulysse) [inline conf/modules/germlinecnvcaller_cohort.config]: why this change?

Comment (Jorisvansteenbrugge) [reply conf/modules/germlinecnvcaller_cohort.config]: @maxulysse Otherwise only the ploidy model and not the calls are published. They may not be strictly necessary, so if you prefer to only publish the model then I'm fine with reverting the change.

Comment (maxulysse) [reply conf/modules/germlinecnvcaller_cohort.config]: I'll leave @ramprasadn to assess that

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 5]
```diff
@@ -167,6 +169,48 @@ workflow GERMLINECNVCALLER_COHORT {
...
+            .collect(flat:false)
+            .set{base_cnv_calls}
+
+        ch_sample_ids.combine(base_cnv_calls)
+            .map{data -> [
+                meta: data[0],
+                calls: data[1],
+                model: data[2],
+                ploidy: data[3]
+            ]}
```
Comment (maxulysse) [inline subworkflows/local/germlinecnvcaller_cohort.nf]:
```suggestion
            ]}
```

File: nextflow.config (hunk 3 above)
Comment (Jorisvansteenbrugge) [reply nextflow.config]: My bad, that was for testing and I forgot to change it back. I changed it back in #[2605105](https://github.com/Jorisvansteenbrugge/createpanelrefs/commit/26051058085cf0eccc8c0b5d0185a9fb21ab0df2)

File: subworkflows/local/germlinecnvcaller_cohort.nf (hunk 5 above)
Comment (maxulysse) [reply subworkflows/local/germlinecnvcaller_cohort.nf]: we missed that bit.

BTW, you can do commit suggestion, or add suggestion to batch and then commit batch on the files tab to commit directly from GitHub

File: modules/nf-core/gatk4/determinegermlinecontigploidy/main.nf [hunk 6]
```diff
@@ -4,7 +4,11 @@ process GATK4_DETERMINEGERMLINECONTIGPLOIDY {
     label 'process_single'
 
     //Conda is not supported at the moment: https://github.com/broadinstitute/gatk/issues/7811
-    container "nf-core/gatk:4.4.0.0" //Biocontainers is missing a package
+    container "docker.io/broadinstitute/gatk:4.5.0.0" //Biocontainers is missing a package
+    //container "quay.io/biocontainers/gatk4:4.4.0.0--py36hdfd78af_0"
+    containerOptions '-u root:root'
+
+
```
Comment (maxulysse) [inline modules/nf-core/gatk4/determinegermlinecontigploidy/main.nf]: I missed that change.

Changes in nf-core modules should happen upstream in the modules repo not on the pipeline one.

File: modules/nf-core/gatk4/determinegermlinecontigploidy/main.nf [hunk 7]
```diff
@@ -39,6 +43,7 @@ process GATK4_DETERMINEGERMLINECONTIGPLOIDY {
         avail_mem = (task.memory.mega*0.8).intValue()
     }
     """
+    #source activate gatk
```
Comment (maxulysse) [inline modules/nf-core/gatk4/determinegermlinecontigploidy/main.nf]: same issue here

File: modules/nf-core/gatk4/germlinecnvcaller/main.nf [hunk 8]
```diff
@@ -4,6 +4,7 @@ process GATK4_GERMLINECNVCALLER {
 
     //Conda is not supported at the moment: https://github.com/broadinstitute/gatk/issues/7811
     container "nf-core/gatk:4.4.0.0" //Biocontainers is missing a package
+    containerOptions "-u root:root"
```
Comment (maxulysse) [inline modules/nf-core/gatk4/germlinecnvcaller/main.nf]: same issue here

File: conf/modules/germlinecnvcaller_cohort.config (hunk 4 above)
Comment (ramprasadn) [reply conf/modules/germlinecnvcaller_cohort.config]: Calls were intentionally not published. As mentioned in another comment, the idea was to only publish the necessary files for running germlinecnvcaller in case mode. Since calls were not necessary for this purpose, it did not seem logical to publish them as a default. Users can always overwrite this in the config. So in short, I would revert this change.

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 9]
```diff
@@ -167,6 +169,48 @@ workflow GERMLINECNVCALLER_COHORT {
...
+        //
+        ch_readcounts_out
+            .map{idx -> idx[1].baseName}
+            .flatten()
+            .map{data -> [id:data]}
+            .set{ch_sample_ids}
+
+        GATK4_GERMLINECNVCALLER.out.cohortcalls
+            .collect()
+            .map{meta, calls1, meta2, calls2 -> [calls1, calls2]}
```
Comment (ramprasadn) [inline subworkflows/local/germlinecnvcaller_cohort.nf:183]: How about? 
```suggestion
            .collect{it[1]}
```

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 10]
```diff
@@ -167,6 +169,48 @@ workflow GERMLINECNVCALLER_COHORT {
...
+            .set{cnv_calls}
+
+        GATK4_GERMLINECNVCALLER.out.cohortmodel
+            .collect()
+            .map{meta, model, meta2, model2 -> [model, model2]}
```
Comment (ramprasadn) [inline subworkflows/local/germlinecnvcaller_cohort.nf:188]: Same as the one above

File: subworkflows/local/germlinecnvcaller_cohort.nf [hunk 11]
```diff
@@ -167,6 +169,48 @@ workflow GERMLINECNVCALLER_COHORT {
...
+            .collect(flat:false)
+            .set{base_cnv_calls}
+
+        ch_sample_ids.combine(base_cnv_calls)
+            .map{data -> [
+                meta: data[0],
+                calls: data[1],
+                model: data[2],
+                ploidy: data[3]
+            ]}
```
Comment (ramprasadn) [inline subworkflows/local/germlinecnvcaller_cohort.nf:210]:
```suggestion
```

**Review by ramprasadn — COMMENTED** [review]
Great job!

In addition to a few comments on the code, I also have a general suggestion. The purpose of adding this subworkflow is to create the necessary files for running germlinecnvcaller in case mode. I have noticed that some users do not run postprocessgermlinecalls (PPGC) because they are only interested in the models this workflow produces. However, I do believe it would be beneficial to have the option to run PPGC as well. What do you think about implementing a new parameter, such as gcnv_run_postprocess, and enclosing your code within a conditional controlled by that parameter to give users control over whether or not they want to run PPGC? I am thinking something like this: https://github.com/nf-core/raredisease/blob/c7a6f389c9b958bf5099c1f48c9b8145d755510f/subworkflows/local/qc_bam.nf#L48-L51

**Comment by ramprasadn** [comment]
Could you please update the gatk modules used in this subworkflow to their latest versions in nf-core/modules, @Jorisvansteenbrugge? Some of recent changes made there should resolve your permission errors (I assume that's why you added containerOptions to change the user ;)).

**Comment by ramprasadn** [comment]
I nearly forgot. Could you check your output from the postprocessgermlinecnvcalls? I think you need to include the [--sample-index](https://gatk.broadinstitute.org/hc/en-us/articles/21905072427035-PostprocessGermlineCNVCalls#--sample-index) parameter to generate results for specific samples, IIRC by default it only process the first sample.

**Comment by Jorisvansteenbrugge** [comment]
@ramprasadn thank you for all your suggestions! I will hopefully be able to work on them some time next week.

**Comment by maxulysse** [comment]
Solved the merge conflicts, sorry if you're getting any issues on your side @Jorisvansteenbrugge
